### PR TITLE
Intrinsify Activator.CreateInstance<T> for value types with no ctor

### PR DESCRIPTION
**Pull request by filipnavara**
Follow up to #17666, fixes https://github.com/mono/mono/issues/16615#issuecomment-527533992

Before:

|                      Method |     Mean |   Error |  StdDev |
|---------------------------- |---------:|--------:|--------:|
| CreateInstanceGenericStruct | 274.8 ns | 0.62 ns | 0.55 ns |
|  CreateInstanceGenericClass | 335.9 ns | 0.61 ns | 0.51 ns |

After:

|                      Method |       Mean |     Error |    StdDev |
|---------------------------- |-----------:|----------:|----------:|
| CreateInstanceGenericStruct |   2.505 ns | 0.0859 ns | 0.1411 ns |
|  CreateInstanceGenericClass | 317.281 ns | 6.3202 ns | 8.8601 ns |

### CODE REVIEW

**Comment by vargaz** [comment]
The netcore failures might be relevant.

File: mono/metadata/class-accessors.c [hunk 1]
```diff
@@ -442,6 +446,37 @@ mono_class_has_dim_conflicts (MonoClass *klass)
...
+mono_method_is_constructor (MonoMethod *method)
+{
+	return ((method->flags & CTOR_REQUIRED_FLAGS) == CTOR_REQUIRED_FLAGS &&
+			!(method->flags & CTOR_INVALID_FLAGS) &&
+			!strcmp (".ctor", method->name));
+}
+
+gboolean
+mono_class_has_default_constructor (MonoClass *klass)
+{
```
Comment (vargaz) [inline mono/metadata/class-accessors.c]: This function should go to class.c, this file is about accessors, which should be fast, and this one isn't.

Comment (filipnavara) [reply mono/metadata/class-accessors.c]: I'll move it.

**Comment by filipnavara** [comment]
> The netcore failures might be relevant.

Looking into it.

File: mono/mini/intrinsics.c [hunk 2]
```diff
@@ -1831,6 +1831,26 @@ mini_emit_inst_for_method (MonoCompile *cfg, MonoMethod *cmethod, MonoMethodSign
...
+		if (!strcmp (cmethod->name, "CreateInstance") &&
+				fsig->param_count == 0 &&
+				method_context != NULL &&
+				method_context->method_inst->type_argc == 1 &&
+				cmethod->is_inflated &&
+				!mini_method_check_context_used (cfg, cmethod)) {
+			MonoClass *arg0 = mono_class_from_mono_type_internal (method_context->method_inst->type_argv [0]);
+			if (m_class_is_valuetype (arg0) && !mono_class_has_default_constructor (arg0, FALSE)) {
+				MONO_INST_NEW (cfg, ins, MONO_CLASS_IS_SIMD (cfg, arg0) ? OP_XZERO : OP_VZERO);
+				ins->dreg = mono_alloc_dreg (cfg, STACK_VTYPE);
```
Comment (EgorBo) [inline mono/mini/intrinsics.c]: I bet OP_XZERO won't compile on arm64

Comment (filipnavara) [reply mono/mini/intrinsics.c]: MONO_CLASS_IS_SIMD will return 0 on ARM

Comment (filipnavara) [reply mono/mini/intrinsics.c]: https://github.com/mono/mono/blob/d274cfbacdb2aeab0184662997882f3d78991e48/mono/mini/mini.h#L308-L329

Comment (EgorBo) [reply mono/mini/intrinsics.c]: Yeah should be ok, I thought OP_XZERO usage is wrapped with if-def usually but's not

**Comment by filipnavara** [comment]
@monojenkins build Linux WebAssembly

**Comment by filipnavara** [comment]
@monojenkins build Windows x64 Cooperative Suspend

**Pull request merged**